src/core.js

```diff
@@ -63,53 +71,106 @@ function createModule( name, testEnvironment ) {
 	return module;
 }
 
-extend( QUnit, {
-	on,
+function processModule( { name, testEnvironment, executeNow, skip = false, todo = false } ) {
+	let module = createModule( { name, testEnvironment, skip, todo } );
 
-	// Call on start of module test to prepend name to all tests
-	module: function( name, testEnvironment, executeNow ) {
-		if ( arguments.length === 2 ) {
-			if ( objectType( testEnvironment ) === "function" ) {
-				executeNow = testEnvironment;
-				testEnvironment = undefined;
-			}
+	// Move any hooks to a 'hooks' object
+	if ( module.testEnvironment ) {
+		module.hooks = {
+			before: module.testEnvironment.before,
+			beforeEach: module.testEnvironment.beforeEach,
+			afterEach: module.testEnvironment.afterEach,
+			after: module.testEnvironment.after
+		};
+
+		delete module.testEnvironment.before;
+		delete module.testEnvironment.beforeEach;
+		delete module.testEnvironment.afterEach;
+		delete module.testEnvironment.after;
+	}
+
+	const moduleFns = {
+		before: setHook( module, "before" ),
+		beforeEach: setHook( module, "beforeEach" ),
+		afterEach: setHook( module, "afterEach" ),
+		after: setHook( module, "after" )
+	};
+
+	const currentModule = config.currentModule;
+	if ( objectType( executeNow ) === "function" ) {
+		moduleStack.push( module );
+		config.currentModule = module;
+		executeNow.call( module.testEnvironment, moduleFns );
+		moduleStack.pop();
+		module = module.parentModule || currentModule;
+	}
+
+	config.currentModule = module;
+}
+
+function module( name, testEnvironment, executeNow ) {
```

Comment:
Can you add a `TODO` comment here noting that we should extract this to a new file? (I think it'll require some refactoring, so I don't want it to be in this PR).